lms/services/oauth_http.py

```diff
@@ -90,11 +100,27 @@ def refresh_access_token(self, token_url, redirect_uri, auth):
         (https://datatracker.ietf.org/doc/html/rfc6749#section-6) to get a new
         access token for the current user and save it to the DB.
 
+        :raise ConflictError: if this token is being refreshed concurrently by another request
         :raise OAuth2TokenError: if we don't have a refresh token for the user
         :raise ExternalRequestError: if the HTTP request fails
         :raise ValidationError: if the server's access token response is invalid
         """
-        refresh_token = self._oauth2_token_service.get(self.service).refresh_token
+        old_token = self._oauth2_token_service.get(self.service)
+
+        # If the "old" token is already current, just return immediately.
+        if (
+            old_token.access_token
+            and datetime.utcnow() - old_token.received_at < timedelta(seconds=30)
+        ):
+            return old_token.access_token
+
+        # Prevent concurrent refresh attempts. If the client gets this, it
+        # should wait briefly and try again, at which point it should find the
+        # refreshed token already available and skip the refresh.
+        try:
+            self._oauth2_token_service.try_lock_for_refresh(self.service)
+        except TryLockError as exc:
+            raise ConflictError() from exc
```

Comment:
Will this add the lock behaviour to _all_ OAuth 2 token refreshes?

Might want to enable it for Canvas Studio admin keys only at first (maybe via a `lock=False` optional argument to this method) as a way to roll it out slowly at first.

Even in the longer-term, I don't think locking is necessary for most refreshes and is a potential source of issues (e.g. Postgres has a global limit on how many locks there can be) so we might not want to use it for all refreshes,

Reply:
> Will this add the lock behaviour to all OAuth 2 token refreshes?

Yes. The issue I see with using it only for Canvas Studio is that service has zero usage right now, and I expect to have very low usage compared with eg. PDFs for the foreseeable future. Using it everywhere may help to reduce unnecessary re-authorizations, but it would also mean that the code gets exercised regularly.

Since these locks are transaction scoped and we can only have a relatively small number of transactions in flight at once (N servers * M workers-per-server) I'm not too worried about hitting limits here.

I do agree that it would make sense to limit locking to Canvas Studio for the initial roll-out though.